### PR TITLE
Feat add @iceworks/spec

### PR DESCRIPTION
**Pull request by fyangstudio**
# @iceworks/spec

Easy to use **eslint**(support TypeScript) / **stylelint** / **prettier** / **commitlint** in [rax](https://rax.js.org/), [ice](https://ice.work/) and react project. And spec means specification.

## Install

```bash
$ npm i --save-dev @iceworks/spec eslint stylelint prettier @commitlint/cli
```

#### 1. Create configuration file

First create a `.eslintrc.js` file. Then edit your config.

##### JavaScript + [rax](https://rax.js.org/), [ice](https://ice.work/) and react

[rules](./src/eslint/ice.js) base on [eslint-config-ali](https://www.npmjs.com/package/eslint-config-ali).

```js
// .eslintrc.js
const { getESLintConfig } = require('@iceworks/spec');

// getESLintConfig(rule: 'rax'|'ice', customConfig?);
module.exports = getESLintConfig('ice');
```

##### TypeScript + [rax](https://rax.js.org/), [ice](https://ice.work/) and react

[rules](./src/eslint/ice-ts.js) base on [eslint-config-ali](https://www.npmjs.com/package/eslint-config-ali).

```js
// .eslintrc.js
const { getESLintConfig } = require('@iceworks/spec');

// getESLintConfig(rule: 'rax-ts'|'ice-ts', customConfig?);
module.exports = getESLintConfig('ice-ts');
```

#### 1. Create configuration file

First create a `.stylelintrc.js` file. Then edit your config.

#### 2. Update config

[rules](./src/stylelint/ice.js) base on [stylelint-config-ali](https://www.npmjs.com/package/stylelint-config-ali)

```js
// .stylelintrc.js
const { getStylelintConfig } = require('@iceworks/spec');

// getStylelintConfig(rule: 'rax'|'ice', customConfig?);
module.exports = getStylelintConfig('ice');
```

#### 1. Create configuration file

First create a `.prettierrc.js` file. Then edit your config.

#### 2. Update config

```js
// .prettierrc.js
const { getPrettierConfig } = require('@iceworks/spec');

// getPrettierConfig(rule: 'rax'|'ice', customConfig?);
module.exports = getPrettierConfig('ice');
```

#### 1. Create configuration file

First create a `.commitlintrc.js` file. Then edit your config.

#### 2. Update config

[rules](./src/commitlint/ice.js) base on [commitlint-config-ali](https://www.npmjs.com/package/commitlint-config-ali)

```js
// .commitlintrc.js
const { getCommitlintConfig } = require('@iceworks/spec');

// getCommitlintConfig(rule: 'rax'|'ice', customConfig?);
module.exports = getCommitlintConfig('ice');
```

### Custom config

```js
// .eslintrc.js
const { getESLintConfig } = require('@iceworks/spec');

// getESLintConfig(rule: 'rax'|'ice', customConfig?);
module.exports = getESLintConfig('rax', {
  // custom config it will merge into main config
  rules: {
    // ...
  },
});
```

### Error: Cannot find module 'eslint-plugin-foo'

Eslint is not yet supported having plugins as dependencies in shareable config. [issue](https://github.com/eslint/eslint/issues/3458). As a temporary solution, you need add the plugin to devDependencies in your project, like `npm i --save-dev eslint-plugin-jsx-a11y`.

Enjoy!

### CODE REVIEW

**Comment by imsobear** [comment]
- best-practices 相关的包应该也放到这个仓库里吧
- `@iceworks/spec` 改成 `@fed/spec` 或者 fed-spec 更好一点
- ice 是不是可以考虑改成 react

File: README.md [hunk 1]
```diff
@@ -1,93 +1,158 @@
...
 ```
 
-### Error: Cannot find module 'eslint-plugin-foo'
+### package.json scripts
+
+Add `scripts` in your `package.json`, example: 
+```json
+"scripts": {
+  "lint": "npm run eslint && npm run stylelint",
+  "eslint": "eslint --cache --ext .js,.jsx,ts,tsx ./",
```
Comment (imsobear) [inline README.md]: .ts .tsx

File: README.md [hunk 2]
```diff
@@ -1,93 +1,158 @@
...
+    // ...
   },
 });
 ```
 
-### Error: Cannot find module 'eslint-plugin-foo'
+### package.json scripts
+
+Add `scripts` in your `package.json`, example: 
+```json
```
Comment (imsobear) [inline README.md:106]: 加个空行

File: README.md [hunk 3]
```diff
@@ -1,93 +1,158 @@
...
+    "pre-push": "npm run lint"
+  }
+}
+```
+
+* `commit-msg`: check commit message
+* `pre-push`: run scripts when push
 
-Eslint is not yet supported having plugins as dependencies in shareable config. [issue](https://github.com/eslint/eslint/issues/3458). As a temporary solution, you need add the plugin to devDependencies in your project, like `npm i --save-dev eslint-plugin-jsx-a11y`.
+## Run Test
```
Comment (imsobear) [inline README.md]: 改成 Develop ？

Comment (imsobear) [reply README.md]: 包含下面这几个，都属于 Develop？

File: packages/spec/package.json [hunk 4]
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "@iceworks/spec",
+  "version": "0.1.1",
```
Comment (imsobear) [inline packages/spec/package.json]: 直接发 1.0 吧

File: packages/spec/package.json [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    "eslint-react": "eslint -c .eslintrc.react.js --ext .js,.jsx ./examples/react/",
+    "eslint-react-ts": "eslint -c .eslintrc.react-ts.js --ext .ts,.tsx ./examples/react-ts/",
+    "eslint-test": "npm run eslint-rax && npm run eslint-rax-ts && npm run eslint-react && npm run eslint-react-ts",
+    "stylelin-test": "stylelint ./**/*.{css,scss}",
+    "test": "npm run eslint-test && npm run stylelin-test",
+    "prepublishOnly": "npm run test"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/ice-lab/iceworks.git"
```
Comment (imsobear) [inline packages/spec/package.json]: 修正下，其他的同理

File: packages/spec/src/deepmerge.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+      newObj[key] = [...(target[key] || []), ...source[key]];
+    } else if (type === '[object Object]') {
+      newObj[key] = { ...(target[key] || {}), ...source[key] };
+    } else {
+      newObj[key] = source[key];
+    }
+  });
+
+  return newObj;
+};
```
Comment (imsobear) [inline packages/spec/src/deepmerge.js]: 这个方法 repeat 的有点多，要不要都依赖 spec 里的？或者单独一个包？

**Comment by imsobear** [comment]
https://github.com/ice-lab/spec/issues/13 这个 issue 和 pr 顺便看下

**Comment by fyangstudio** [comment]
> #13 这个 issue 和 pr 顺便看下

此分支已修复

File: packages/spec/src/deepmerge.js (hunk 6 above)
Comment (fyangstudio) [reply packages/spec/src/deepmerge.js]: 已使用 const {deepmerge} = require('@iceworks/spec'); 替换

**Review by imsobear — APPROVED** [review]

**Pull request merged**